### ISSUE TITLE
Prevent formatting links as markdown links

### ISSUE THREAD

**Issue by ekzyis** [stackernews/stacker.news#323]
Currently, users can create open redirects to potentially malicious links.

For example I can write [https://stacker.news/items/194732](https://stacker.news/items/194731) which actually redirects to https://stacker.news/items/194731
(interestingly, this also works in Github).

At first, I thought a open redirect warning would be enough but since we use a lot of links to external sites on SN, I think this would be detrimental to UX.

So I think the better solution is to check if the thing inside `[]` is already a link and then prevent using the markdown link syntax (`[]()`)

**Comment by IonIsle** [comment]
Hey i'm about to for the repo and start working on this @huumn or another dev, could you point me in the right direction of the flow and the files I'll need to touch? I can stumble through it, but with a bit of guidance I could probably get it wrapped up pretty quickly. Just getting my sea legs in your software :D

**Comment by ekzyis** [comment]
For this ticket, I would guess that the important files are `components/text.js` since that's where our Markdown input is defined (it uses [`react-markdown`](https://github.com/remarkjs/react-markdown)) and `lib/md.js` for other markdown stuff

**Comment by ed-kung** [comment]
I've been looking at this issue and it looks like a partial solution is already implemented in `lib/rehype-sn.js`

https://github.com/stackernews/stacker.news/blob/a4cce7afed8813097dfc020aec16401ae0991ee3/lib/rehype-sn.js#L85C1-L89C1

https://github.com/stackernews/stacker.news/blob/a4cce7afed8813097dfc020aec16401ae0991ee3/lib/rehype-sn.js#L253C1-L265C1

Basically, if the link text is URL-like (numbers letters and dots), and the resolved origin is different than the href, then the link is labeled as misleading, and the link text is simply replaced by href.

The issue here is that `isMisleadingLink` won't check for an origin match if `text` contains characters like `:` and `/`, which may occur in many urls, like "https://stacker.news/items/1234"

Thus: 
- `[https://bing.com](https://google.com)` resolves to https://bing.com for the text and https://google.com for the href <-- `isMisleadingLink` didn't work
- But `[bing.com](https://google.com)` resolves to https://google.com for both the text and the href <-- `isMisleadingLink` worked
- `[https://stacker.news/items/1235](https://stacker.news/items/1234])` resolves to https://stacker.news/items/1235 for the text and https://stacker.news/items/1234 for the link <-- `isMisleadingLink` didn't work

The solution seems to be that `isMisleadingLink` should also check whether `text` contains a valid URL and if so, check whether it's the same link as `href`, and if not, replace the link text with href.

If this seems like a reasonable solution, I can work on changing `isMisleadingLink` to do the above.

**Comment by ekzyis** [comment]
Ohh, good catch! I agree that `[https://bing.com](https://google.com)` should render as https://google.com.

However, contrary to the issue description with the stacker.news link example, I think if the origin is the same, we shouldn't handle it as misleading since it can't be used for phishing attacks as mentioned [here](https://en.wikipedia.org/wiki/Phishing#Link_manipulation):

> Another tactic is to make the displayed text for a link appear trustworthy, while the actual link goes to the phisher's site.

What do you think?

> If this seems like a reasonable solution, I can work on changing `isMisleadingLink` to do the above.

Sure!

**Comment by ed-kung** [comment]
Makes sense. It won't solve the misdirection from one SN item to another, as in the OP, but that seems to be a much smaller issue than misdirecting to an entirely different domain. And there might be legitimate use cases for something like `[https://stacker.news](https://stacker.news/recent)` Will try to get a PR up soon.
